pkg/volume/scaleio/sio_mgr.go

```diff
@@ -0,0 +1,266 @@
+	configs[confKey.storageMode] = defaultString(configs[confKey.storageMode], "ThinProvisioned")
+
+	mgr := &sioMgr{configData: configs}
+	return mgr, nil
+}
+
+// getClient safely returns an sioInterface
+func (m *sioMgr) getClient() (sioInterface, error) {
+	if m.client == nil {
+		glog.V(4).Info(log("creating scaleio client"))
+		configs := m.configData
+		username := configs[confKey.username]
+		password := configs[confKey.password]
+		gateway := configs[confKey.gateway]
+		b, _ := strconv.ParseBool(configs[confKey.sslEnabled])
+		certsEnabled := b
+
+		glog.V(4).Info(log("creating new client for gateway %s", gateway))
+		client, err := newSioClient(gateway, username, password, certsEnabled)
+		if err != nil {
+			glog.Error(log("failed to create scaleio client: %v", err))
+			return nil, err
+		}
+
+		client.sysName = configs[confKey.system]
+		client.pdName = configs[confKey.protectionDomain]
+		client.spName = configs[confKey.storagePool]
+		client.sdcPath = configs[confKey.sdcRootPath]
+		client.provisionMode = configs[confKey.storageMode]
+
+		m.client = client
+
+		glog.V(4).Info(log("client created successfully [gateway=%s]", gateway))
+	}
+	return m.client, nil
+}
+
+// CreateVolume creates a new ScaleIO volume
+func (m *sioMgr) CreateVolume(volName string, sizeGB int64) (*siotypes.Volume, error) {
+	client, err := m.getClient()
+	if err != nil {
+		return nil, err
+	}
+
+	glog.V(4).Infof("scaleio: creating volume %s", volName)
+	vol, err := client.CreateVolume(volName, sizeGB)
+	if err != nil {
+		glog.V(4).Infof("scaleio: failed creating volume %s: %v", volName, err)
+		return nil, err
+	}
+	glog.V(4).Infof("scaleio: created volume %s successfully", volName)
+	return vol, nil
+}
+
+// AttachVolume maps a ScaleIO volume to the running node
+func (m *sioMgr) AttachVolume(volName string) (string, error) {
+	client, err := m.getClient()
+	if err != nil {
+		glog.Error(log("attach volume failed: %v", err))
+		return "", err
+	}
+
+	glog.V(4).Infoln(log("attaching volume %s", volName))
+	iid, err := client.IID()
+	if err != nil {
+		glog.Error(log("failed to get instanceID"))
+		return "", err
+	}
+	glog.V(4).Infof(log("attaching volume %s to host instance %s", volName, iid))
+
+	devs, err := client.Devs()
+	if err != nil {
+		return "", err
+	}
+
+	vol, err := client.FindVolume(volName)
+	if err != nil {
+		glog.Error(log("failed to find volume %s: %v", volName, err))
+		return "", err
+	}
+
+	// handle vol if already attached
+	if len(vol.MappedSdcInfo) > 0 {
+		glog.V(4).Info(log("vol %s has attachments", volName))
+
+		// if vol attached to current instance, done
+		if m.isSdcMappedToVol(iid, vol) {
+			glog.V(4).Info(log("skippping attachment, volume %s already attached to sdc %s", volName, iid))
+			return devs[vol.ID], nil
+		}
+
+		// preempt attachment if vol attached to a different sdc
+		glog.V(4).Info(log("preemting attachment for vol %s", volName))
+		if err := client.DetachVolume(vol.ID); err != nil {
```

Comment:
not sure if preempting is a good idea. What if the volume is still being used? would it cause data loss?